torch/csrc/distributed/rpc/RpcAgent.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <torch/csrc/distributed/rpc/FutureMessage.h>
+#include <torch/csrc/distributed/rpc/Message.h>
+#include <torch/csrc/distributed/rpc/rpc_headers.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+
+// RpcAgent is the base class for sending and receiving RPC messages. It
+// provides a unified ``send`` API for both request and response messages, and
+// will invoke the given ``RequestCallback`` to process received requests. It
+// should immediately become ready to serve request and accept response after
+// construction.
+class RpcAgent;
+
+// RpcAgent implementation should invoke ``RequestCallback`` to process received
+// requests. There is no restriction on the implementation's threading model.
+// This function takes the name of the request sender, the an rvalue reference
+// of the Message object, and a reference to the RpcAgent itself. Having a
+// reference to the RpcAgent allows the ``RequestCallback`` implementation to
+// be both stateless and non-blocking. It may enqueue the message and the
+// RpcAgent reference, and use a different set of threads to process them later.
+using RequestCallback = std::function<void(std::string, Message&&, RpcAgent&)>;
+
+class RpcAgent {
+ public:
+  // The ``workerName`` is the globally unique name for this RpcAgent. It is up
+  // to the RpcAgent implementation to determine how to resolve names.
+  // The ``RequestCallback`` will be invoked to handle received requests. This
+  // RpcAgent base class makes no assumption on the thread-safeness of the
+  // ``RequestCallback``. RpcAgent implementations need to make sure that its
+  // threading model conform to ``RequestCallback``'s requirement.
+  RpcAgent(std::string workerName, RequestCallback cb);
+
+  virtual ~RpcAgent() noexcept(false);
```

Comment:
Omg, please don't do throwing destructors...

Reply:
This is a temporary solution to exit `recvAnysource`. Currently, we don't have a way to gracefully exit `recvAnysource` (and hence the `listenLoop` thread). What we did is requiring applications to call `RpcAgent::shutdown` which will explicitly send a `SHUTDOWN` message to a neighbor `ProcessGroupAgent` instance. We cannot do this in the destructor as it is not safe to call virtual methods in destructors. The `noexcept(false)` here is to raise the alarm when the application forgets to call `shutdown`. 

Or will it be better if we just show a warning instead of throwing in this destructor?